src/Voron/Data/CompactTrees/CompactTree.Dictionaries.cs

```diff
@@ -0,0 +1,106 @@
+// compress the whole tree in the most efficient way (if needed).
+//  - Pro: There is only a single active dictionary that everybody uses,
+//  - Pro: The performance burden could be paid as an idle task and happen at the moment of splitting,
+//  - Pro: The pages could be upgraded over time when performing writes on a particular page to amortize the IO cost,
+//  - Pro: Random selection or fixed budget would do the trick of adaptively upgrading the tree compression over time,
+//  - Cons: Requires to have a task to perform the dictionary creation outside of the insertion process,
+//  - Cons: Compression ratio will not be THAT tight per single page,
+//  - Cons: We may need larger dictionaries to accomodate better compression ratios. 
+//
+//  In Voron we use global scopes dictionaries because they have a cleaner implementation with less intrusive processes
+//  where it really matters (writing). 
+
+unsafe partial class CompactTree
+{
+    private struct FullDictionaryKeyScanner : IReadOnlySpanEnumerator
+    {
+        private readonly CompactTree _tree;
+        private Iterator _iterator;
+
+        public FullDictionaryKeyScanner(CompactTree tree)
+        {
+            _tree = tree;
+            _iterator = _tree.Iterate();
+        }
+
+        public bool MoveNext(out ReadOnlySpan<byte> result)
+        {            
+            bool operationResult = _iterator.MoveNext(out Span<byte> resultSlice, out long _);
+            result = resultSlice;
+            return operationResult;
+        }
+
+        public void Reset()
+        {
+            _iterator = _tree.Iterate();
+        }
+    }
+
+    /// <summary>
+    /// We will try to improve the dictionary by scanning the whole tree and using the data. This may be costly and alternative and probably
+    /// less effective solutions may need to be developed for when this method becomes prohibitive. 
+    /// </summary>    
+    public bool TryImproveDictionaryByFullScanning()
+    {
+        Debug.Assert(_llt.Flags == TransactionFlags.ReadWrite);
+
+        if (_state.NextTrainAt < _state.NumberOfEntries)
+            return false;
+
+        // We will try to improve the dictionary by scanning the whole tree over a maximum budget.    
+        // We will do this by randomly selecting a page and then randomly selecting a key from that page.
+        var currentDictionary = GetEncodingDictionary(_state.TreeDictionaryId);
+        var newDictionary = PersistentDictionary.CreateIfBetter(_llt, new FullDictionaryKeyScanner(this), new FullDictionaryKeyScanner(this), currentDictionary);
+
+        // If the new dictionary is actually better, then update the current dictionary at the tree level.    
+        if (currentDictionary != newDictionary)
+            _state.TreeDictionaryId = newDictionary.PageNumber;
+
+        // We will update the number of entries regardless if we updated the current dictionary or not. 
+        _state.NumberOfEntries = (int)(_state.NumberOfEntries * 1.5);
```

Comment:
Isn't that the number of entries in the _tree_, not the dictionary.
Also, what about overflow here? We should be prepared from > 2B entries in the tree

Reply:
Good point that is a long, either way something is off there. `NumberOfEntries` is not the item to update. There must have been a mixed up during cleanup.